index.rst

```diff
@@ -1,57 +1,451 @@
+* It should persist files in a human-readable form, where practical, as a debugging aid.
+
+We do not have any expectation that we should be able to easily change persistence frameworks in the future.
+
+Option: Avro
+------------
+
+.. _Avro: https://avro.apache.org/
+
+`Avro`_ is a table-like persistence library provided by Apache.
+It defines persistence formats in terms of schemas, which may be composed (e.g., the user can declare that a :class:`lsst.geom.Box2I` is stored as a pair of :class:`lsst.geom.Point2I`, as long as there is a persisted form for :class:`~lsst.geom.Point2I`).
+In C++, the schemas are typically compiled into proxy classes, whose data are then accessed using object member syntax.
+In Python, no code generation is required, and data are represented using :class:`dict`-like objects.
+Avro uses a schema definition language based on JSON, but as the notation is similar in spirit to :class:`lsst.pipe.base.PipelineTaskConnections`, it should not be a major conceptual burden on LSST developers.
+
+Because of its record-based approach, Avro is best-suited for bulk data: one could store a :class:`~lsst.afw.table.Catalog` very naturally using Avro, but a free compound object like an :class:`~lsst.afw.image.Exposure` would have some overhead (essentially, its persisted form would be a collection of one-row tables).
+
+Avro meets many, but not all, of our criteria:
+
+* It uses the Apache 2.0 license.
+* It has built-in support for both C++ and Python, although the Python API is much better.
+* It has built-in support for JSON, as well as a proprietary format. We can, in principle, support additional formats using stream I/O, but the interfaces do not appear to have been designed for third-party implementations.
+* It supports schema versioning.
+* It may be possible to depersist ``lsst.afw.table.io`` files by treating it as a custom format.
+* It natively supports arrays of arbitrary type, including numeric primitives.
+* It does not natively handle polymorphism, though we could emulate it by storing an explicit class name (much like ``lsst.afw.table.io`` does).
+  Avro makes heavy use of dynamic typing on depersistence, so differences in the persisted form among related classes are not a problem.
+* Its tables are row-oriented, so it does not have any special ability to partially read objects.
+* It does not natively store object references, though we could emulate them by introducing a unique object ID data type.
+* In Python, the persisted form is a :class:`dict` from field names to field values, which is reasonably human-readable.
+
+Option: cereal
+--------------
+
+.. _cereal: https://uscilab.github.io/cereal/
+
+.. _Boost Serialization: https://www.boost.org/doc/libs/release/libs/serialization/
+
+`cereal`_ is a C++ persistence library provided by USC.
```

Comment:
https://www.acronymfinder.com/USC.html :-)

It's not hard to figure out which you mean, but if it's worth mentioning at all, I think it's worth spelling out in full.